// @mux/mux-node/dist/video/resources/liveStreams.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/video/resources/liveStreams.ts
var liveStreams_exports = {};
__export(liveStreams_exports, {
  LiveStreams: () => LiveStreams
});
module.exports = __toCommonJS(liveStreams_exports);

// src/base.ts
var import_axios = __toESM(require("axios"));
var import_events = __toESM(require("events"));

// src/version.ts
var VERSION = "7.3.0";

// src/base.ts
var Base = class extends import_events.default {
  constructor(tokenIdOrOptionsOrBase, tokenSecret, config) {
    var _a, _b, _c, _d, _e, _f, _g;
    super();
    if (tokenIdOrOptionsOrBase instanceof Base) {
      this.config = tokenIdOrOptionsOrBase._config;
      this._tokenId = tokenIdOrOptionsOrBase._tokenId;
      this._tokenSecret = tokenIdOrOptionsOrBase._tokenSecret;
      this.http = tokenIdOrOptionsOrBase.http;
    } else {
      if (typeof tokenIdOrOptionsOrBase === "object" && !(tokenIdOrOptionsOrBase instanceof Base)) {
        this.config = tokenIdOrOptionsOrBase;
        this.tokenId = void 0;
        this.tokenSecret = void 0;
      } else {
        this.tokenId = tokenIdOrOptionsOrBase;
        this.tokenSecret = tokenSecret;
        this.config = config ?? {};
      }
      const request = {
        baseURL: this.config.baseUrl,
        headers: {
          "User-Agent": `Mux Node | ${VERSION}`,
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        withCredentials: false,
        auth: {
          username: this._tokenId,
          password: this._tokenSecret
        }
      };
      if ((_a = this.config.platform) == null ? void 0 : _a.name) {
        if ((_c = (_b = this.config.platform) == null ? void 0 : _b.name) == null ? void 0 : _c.includes("|")) {
          throw new Error('Platform name cannot contain a "|" value.');
        }
        if ((_e = (_d = this.config.platform) == null ? void 0 : _d.version) == null ? void 0 : _e.includes("|")) {
          throw new Error('Platform version cannot contain a "|" value.');
        }
        request.headers["x-source-platform"] = `${(_f = this.config.platform) == null ? void 0 : _f.name} | ${(_g = this.config.platform) == null ? void 0 : _g.version}`;
      }
      this.http = import_axios.default.create(request);
      this.http.interceptors.request.use((req) => {
        this.emit("request", req);
        return req;
      });
      this.http.interceptors.response.use((res) => {
        this.emit("response", res);
        if (res.config.url && this.isVideoUrl(res.config.url)) {
          return res.data && res.data.data;
        }
        return res.data;
      }, (errorRes) => Promise.reject(errorRes.response && errorRes.response.data.error || errorRes));
    }
  }
  isVideoUrl(url) {
    return url.startsWith(`/video/v1/`);
  }
  set config(options) {
    this._config = {
      baseUrl: "https://api.mux.com",
      ...options
    };
  }
  get config() {
    return this._config;
  }
  set tokenId(token) {
    const v = token || process.env.MUX_TOKEN_ID;
    if (!v || v.length === 0) {
      throw new Error("API Access Token must be provided.");
    }
    this._tokenId = v;
  }
  get tokenId() {
    return this._tokenId;
  }
  set tokenSecret(secret) {
    const v = secret || process.env.MUX_TOKEN_SECRET;
    if (!v || v.length === 0) {
      throw new Error("API secret key must be provided");
    }
    this._tokenSecret = v;
  }
  get tokenSecret() {
    return this._tokenSecret;
  }
};

// src/video/resources/liveStreams.ts
var PATH = "/video/v1/live-streams";
var buildBasePath = (liveStreamId) => `${PATH}/${liveStreamId}`;
var LiveStreams = class extends Base {
  create(params) {
    return this.http.post(PATH, params);
  }
  async update(liveStreamId, params) {
    if (!liveStreamId || !params) {
      throw new Error("assetId and params are required.");
    }
    return this.http.patch(buildBasePath(liveStreamId), params);
  }
  del(liveStreamId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to delete a live stream"));
    }
    return this.http.delete(buildBasePath(liveStreamId));
  }
  get(liveStreamId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to get a live stream"));
    }
    return this.http.get(buildBasePath(liveStreamId));
  }
  list(params) {
    return this.http.get(PATH, { params });
  }
  signalComplete(liveStreamId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to signal a stream is complete"));
    }
    return this.http.put(`${buildBasePath(liveStreamId)}/complete`);
  }
  resetStreamKey(liveStreamId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to reset a live stream key"));
    }
    return this.http.post(`${buildBasePath(liveStreamId)}/reset-stream-key`);
  }
  createPlaybackId(liveStreamId, params) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to create a live stream playback ID"));
    }
    if (!params) {
      return Promise.reject(new Error("A playback policy is required to create a live stream playback ID"));
    }
    return this.http.post(`${buildBasePath(liveStreamId)}/playback-ids`, params);
  }
  deletePlaybackId(liveStreamId, playbackId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to delete a live stream playback ID"));
    }
    if (!playbackId) {
      return Promise.reject(new Error("A live stream playback ID is required to delete a live stream playback ID"));
    }
    return this.http.delete(`${buildBasePath(liveStreamId)}/playback-ids/${playbackId}`);
  }
  playbackId(liveStreamId, playbackId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required"));
    }
    if (!playbackId) {
      return Promise.reject(new Error("A playback ID is required"));
    }
    return this.http.get(`${buildBasePath(liveStreamId)}/playback-ids/${playbackId}`);
  }
  createSimulcastTarget(liveStreamId, params) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to create a simulcast target"));
    }
    if (!(params && params.url)) {
      return Promise.reject(new Error("A url is required to create a simulcast target"));
    }
    return this.http.post(`${buildBasePath(liveStreamId)}/simulcast-targets`, params);
  }
  getSimulcastTarget(liveStreamId, simulcastTargetId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to get a simulcast target"));
    }
    if (!simulcastTargetId) {
      return Promise.reject(new Error("A simulcast target ID is required to get a simulcast target"));
    }
    return this.http.get(`${buildBasePath(liveStreamId)}/simulcast-targets/${simulcastTargetId}`);
  }
  deleteSimulcastTarget(liveStreamId, simulcastTargetId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to delete a simulcast target"));
    }
    if (!simulcastTargetId) {
      return Promise.reject(new Error("A simulcast target ID is required to delete a simulcast target"));
    }
    return this.http.delete(`${buildBasePath(liveStreamId)}/simulcast-targets/${simulcastTargetId}`);
  }
  async updateEmbeddedSubtitles(liveStreamId, params) {
    if (!liveStreamId || !params) {
      throw new Error("liveStreamId and params are required.");
    }
    return this.http.put(`${buildBasePath(liveStreamId)}/embedded-subtitles`, params);
  }
  async updateGeneratedSubtitles(liveStreamId, params) {
    if (!liveStreamId || !params) {
      throw new Error("liveStreamId and params are required.");
    }
    return this.http.put(`${buildBasePath(liveStreamId)}/generated-subtitles`, params);
  }
  disable(liveStreamId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to disable a live stream"));
    }
    return this.http.put(`${buildBasePath(liveStreamId)}/disable`);
  }
  enable(liveStreamId) {
    if (!liveStreamId) {
      return Promise.reject(new Error("A live stream ID is required to enable a live stream"));
    }
    return this.http.put(`${buildBasePath(liveStreamId)}/enable`);
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LiveStreams
});
/*!
 * Mux Live Streams
 * Copyright(c) 2018 Mux Inc.
 */
module.exports = module.exports.default || module.exports;
